### DISCUSSION TITLE
PresentationTreeDataProvider.pagingSize

### DISCUSSION THREAD

**Discussion by paprophet** [iTwin/itwinjs-core#857] [Q&A] [answered]
when open  bim file,the brower console this:

Warning |presentation-components.Hierarchy| PresentationTreeDataProvider.pagingSize doesn't match pageOptions in PresentationTreeDataProvider.getNodes call.
        Make sure you set PresentationTreeDataProvider.pagingSize to avoid excessive backend requests.

but in  code,no use the PresentationTreeDataProvider 
Presentation.initialize method must set pagingSize ??

initPromises.push(Presentation.initialize({
      activeLocale: IModelApp.i18n.languageList()[0],
    }));

**Comment by grigasp** [comment, marked as answer]
The warning originates from right here: https://github.com/imodeljs/imodeljs/blob/5b8fd1fd7e8501ad0a252fd9d7e79c7d6a620821/presentation/components/src/presentation-components/tree/DataProvider.ts#L146-L148

It is possible that you're using some component that uses this provider and doesn't set the `pagingSize`. Putting the breakpoint at those lines would help identify what that is.

**Reply by paprophet** [reply]
i find the error
in CategoryModelTree.tsx, CategoryModelTree._getTree() method,and  pageSize parameter 
when open some bim file,console this error, so i want what value set  to this pageSize  is right?100?200?

**Reply by grigasp** [reply]
The `CategoryModelTree._getTree()` in our repo is in `ModelSelectorTree.tsx` source file and you say it's in `CategoryModelTree.tsx` - do you have a copy of that component locally? 

The log message warns you about different page sizes being set on the data provider compared to page size being used when requesting nodes from it. When they're the same, we can save one trip to the backend, so we emit the warning to tell that. I looked at the [CategoryModelTree](https://github.com/imodeljs/imodeljs/blob/de526f2045277089290f3f615a09f2c146552847/ui/framework/src/ui-framework/pickers/ModelSelector/ModelSelectorTree.tsx#L28) and it looks like it's set up correctly:
- It uses [ModelSelectorDataProvider](https://github.com/imodeljs/imodeljs/blob/de526f2045277089290f3f615a09f2c146552847/ui/framework/src/ui-framework/pickers/ModelSelector/ModelSelectorDefinitions.tsx#L114) which [sets the pagingSize to 5](https://github.com/imodeljs/imodeljs/blob/de526f2045277089290f3f615a09f2c146552847/ui/framework/src/ui-framework/pickers/ModelSelector/ModelSelectorDefinitions.tsx#L120).
- It [passes page size of 5 to the underlying tree component](https://github.com/imodeljs/imodeljs/blob/de526f2045277089290f3f615a09f2c146552847/ui/framework/src/ui-framework/pickers/ModelSelector/ModelSelectorTree.tsx#L500) - this parameter is used when requesting nodes.

I should also note that `CategoryModelTree` is deprecated and we're planning to remove it on next major version release. Recommended replacements are [ModelsTree](https://github.com/imodeljs/imodeljs/blob/de526f2045277089290f3f615a09f2c146552847/ui/framework/src/ui-framework/imodel-components/models-tree/ModelsTree.tsx#L98) and [CategoryTree](https://github.com/imodeljs/imodeljs/blob/de526f2045277089290f3f615a09f2c146552847/ui/framework/src/ui-framework/imodel-components/category-tree/CategoriesTree.tsx#L79).

**Reply by paprophet** [reply]
i use the ModelSelectorWidgetControl  and this include CategoryModelTree ,so have this error

**Reply by grigasp** [reply]
`ModelSelectorWidgetControl` [is `@internal`](https://github.com/imodeljs/imodeljs/blob/de526f2045277089290f3f615a09f2c146552847/ui/framework/src/ui-framework/pickers/ModelSelector/ModelSelector.tsx#L28) and should not be used outside of `imodeljs` repository.
